File: back/src/main/java/com/back/domain/wish/controller/ApiV1WishController.java
package com.back.domain.wish.controller;

import com.back.domain.book.dto.BookDto;
import com.back.domain.book.dto.BookWithTagDto;
import com.back.domain.book.entity.Book;
import com.back.domain.book.repository.BookRepository;
import com.back.domain.book.service.BookService;
import com.back.domain.member.entity.Member;
import com.back.domain.member.service.MemberService;
import com.back.domain.tag.service.TagService;
import com.back.domain.wish.service.WishService;
import com.back.global.rq.Rq;
import com.back.global.rsData.RsData;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/wishes")
@RequiredArgsConstructor
@Transactional(readOnly = true)
@Tag(name = "ApiV1WishController", description = "API 찜 컨트롤러")
@SecurityRequirement(name = "bearerAuth")
public class ApiV1WishController {

    private final WishService wishService;
    private final MemberService memberService;
    private final BookService bookService;

    private final Rq rq;
    private final TagService tagService;

    @GetMapping("/mine")
    @Operation(summary = "내 찜 목록 조회")
    public List<BookWithTagDto> getWishes() {
        Member actor = memberService.findById(rq.getActor().getId());

        return wishService
                .findByMember(actor)
                .stream()
                .map(wish -> new BookWithTagDto(wish.getBook(), bookService.getBookTags(wish.getBook())))
                .toList();
    }

    @PostMapping("/book/{id}")
    @Operation(summary = "찜 목록 추가")
    @Transactional
    public RsData<Void> addWish(
            @PathVariable @Valid long id
    ) {

        Book book = bookService.getPureBook(id);

        wishService.addWish(
                rq.getActorFromDb(),
                book);

        return new RsData<>(
                "201-1",
                "찜 추가 성공"
        );
    }

    @DeleteMapping("/book/{id}")
    @Transactional
    @Operation(summary = "찜 목록 삭제")
    public RsData<Void> deleteWish(
            @PathVariable @Valid long id
    ) {

        Book book = bookService.getPureBook(id);

        wishService.deleteWish(
                rq.getActorFromDb(),
                book);

        return new RsData<>(
                "200-1",
                "찜 삭제 성공"
        );
    }

}

File: back/src/main/kotlin/com/back/domain/wish/controller/ApiV1WishController.kt
package com.back.domain.wish.controller

import com.back.domain.book.dto.BookWithTagDto
import com.back.domain.book.service.BookService
import com.back.domain.wish.service.WishService
import com.back.global.rq.Rq
import com.back.global.rsData.RsData
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/wishes")
@Tag(name = "ApiV1WishController", description = "API 찜 컨트롤러")
@SecurityRequirement(name = "bearerAuth")
@Transactional(readOnly = true)
class ApiV1WishController(
    private val wishService: WishService,
    private val bookService: BookService,
    private val rq: Rq
) {

    @GetMapping("/mine")
    @Operation(summary = "내 찜 목록 조회")
    fun getWishes(): List<BookWithTagDto> {
        val actor = rq.actorFromDb

        return wishService
            .getWishesByMember(actor)
            .map { BookWithTagDto(it.book, bookService.getBookTags(it.book)) }
    }

    @PostMapping("/book/{id}")
    @Operation(summary = "찜 목록 추가")
    @Transactional
    fun addWish(
        @PathVariable @Valid id: Long
    ): RsData<Void> {
        val book = bookService.getBook(id)

        val actor = rq.actorFromDb
        wishService.addWish(actor, book)

        return RsData("201-1", "찜 생성을 성공했습니다.")
    }

    @DeleteMapping("/book/{id}")
    @Operation(summary = "찜 목록 삭제 (구형)")
    @Transactional
    fun oldDeleteWish(
        @PathVariable @Valid id: Long
    ): RsData<Void> {
        val actor = rq.actorFromDb
        val book = bookService.getBookById(id)
        wishService.oldDeleteWish(actor, book)

        return RsData("200-1", "찜 삭제를 성공했습니다.")
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "찜 목록 삭제")
    @Transactional
    fun deleteWish(
        @PathVariable @Valid id: Long
    ): RsData<Void> {
        val actor = rq.actorFromDb
        wishService.deleteWish(actor, id)

        return RsData("200-1", "찜 삭제를 성공했습니다.")
    }
}
